src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ExportedMethodsWriter.cs

```diff
@@ -0,0 +1,97 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+
+using ILCompiler.DependencyAnalysisFramework;
+using Internal.TypeSystem;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    internal class ExportedMethodsWriter
+    {
+        private string _exportsFilePath;
+        private string _moduleName;
+        private TargetDetails _targetDetails;
+        private static ModuleDesc _systemModule;
+
+        private ExportedMethodsWriter(string outputFile, NodeFactory factory)
+        {
+            string directory = Path.GetDirectoryName(outputFile);
+            string filename = Path.GetFileNameWithoutExtension(outputFile);
+
+            _targetDetails = factory.Target;
+            _systemModule = factory.TypeSystemContext.SystemModule;
+            _moduleName = filename;
+            _exportsFilePath = Path.Combine(directory, filename + GetExportsFileExtenstion());
+        }
+
+        public static void EmitExportedMethods(string outputFile, NodeFactory factory)
+        {
+            // Only generate an export file if we're in native library mode
+            // marked by the presense of __managed__Startup method
+            if (!factory.NodeAliases.Values.Contains("__managed__Startup"))
+                return;
+
+            var nativeCallables = factory.NodeAliases.Where(n => n.Key is IMethodNode)
+                                    .Where(n => (n.Key as IMethodNode).Method.IsNativeCallable)
+                                    .Where(n => !IsInternalExport((n.Key as IMethodNode).Method));
+
+            var exportNames = nativeCallables.Select(n => n.Value);
+            new ExportedMethodsWriter(outputFile, factory).WriteExportedMethodsToFile(exportNames);
+        }
+
+        private string GetExportsFileExtenstion()
+        {
+            if (_targetDetails.OperatingSystem == TargetOS.Windows)
```

Comment:
The file name should be an argument on the command line.

Reply:
Won't this be an extra cognitive load for the developer? I was thinking that just having everything done automatically might be better, experience wise

Reply:
I meant it should be argument on ilc.exe command and passed in by the msbuild script. The msbuild script is computing the name already and passing it to the linker, so it can pass it to ilc.exe as well.

Most developers won't invoke ilc.exe directly, so they won't really notice this argument.

Reply:
Gotcha